### ISSUE TITLE
Performance fixes for card template when showing many results on one page

### ISSUE THREAD

**Issue by jhholm** [microsoft-search/pnp-modern-search#2972]
**Performance woes when using card template and e.g. 100 results**
Page loads and pagination changes can be quite slow, and especially on edit mode even on a rather recent PC.

**Description of use case**
Quite common use case is to use PnP Modern Search as a navigation page. E.g to show all project sites that are available. When using the handlebars card template and showing e.g. 50 to 100 results on a page, the generated HTML becomes quite massive. This makes the page load quite slow on lower end systems and mobile. This is probably because each web component (pnp-documentcard) contains a data-context attribute that has all the search result items and some additional duplicated data that are present in other attributes (e.g. theme).

**Examples:**
1. Showing 100 results, default selected properties, compact card template: 320 000 characters per web component => ~32 million characters of HTML for 100 items
2. Removing all other result items from data-context with same scenario in 1 => 26 000 characters per item  => 260 thousand characters of HTML for 100 items

**Describe the solution you would recommend**
I think it might be feasible to refactor what is provided to data-context or refactor the template somewhat. Should this actually be a discussion instead of an issue? 

**I am willing to do a pull request on this, but would actually want to get some feedback what kind of change seem most feasible.**

**Additional things to consider**
I guess it might also be worth to take a look if the images could be lazily loaded when they appear on viewport as that might also factor on the performance.

**Comment by wobba** [comment]
Feel free to propose an optimization :)

**Comment by jhholm** [comment]
I tried to understand the current implementation and come up with a workaround that would not create any breaking changes. I think the context object contains currently duplicate data, as theme is usually provided as another attribute, but it's still used in the template processing.

**data-context** seems to be used only by the following web components
* pnp-documentcard
* pnp-detailslist
* pnp-persona
* pnp-slider

=> Thus JSONStringify @root is only used in those places

**pnp-documentcard** and **pnp-persona** use **ITemplateService processFieldsConfiguration** to render the handlebars template
https://github.com/microsoft-search/pnp-modern-search/blob/cda9ca9fc0ccd3068f1b64ab8b68e2a5c9d4a516/search-parts/src/components/DocumentCardComponent.tsx#L113
https://github.com/microsoft-search/pnp-modern-search/blob/cda9ca9fc0ccd3068f1b64ab8b68e2a5c9d4a516/search-parts/src/components/PersonaComponent.tsx#L100

The following properties are explicitly used in both cases
https://github.com/microsoft-search/pnp-modern-search/blob/cda9ca9fc0ccd3068f1b64ab8b68e2a5c9d4a516/search-parts/src/services/templateService/TemplateService.ts#L373-L384


**pnp-detailslist** has an internal handlebars templating function **this._processHandleBarsExprValue** which uses the same fields from context
https://github.com/microsoft-search/pnp-modern-search/blob/cda9ca9fc0ccd3068f1b64ab8b68e2a5c9d4a516/search-parts/src/components/DetailsListComponent.tsx#L597-L613

Additionally **paging** is used outside of the processing function
https://github.com/microsoft-search/pnp-modern-search/blob/cda9ca9fc0ccd3068f1b64ab8b68e2a5c9d4a516/search-parts/src/components/DetailsListComponent.tsx#L240-L248


**pnp-slider** also has an internal handlebars templating function which just destructures the whole context object, so all properties are used
https://github.com/microsoft-search/pnp-modern-search/blob/cda9ca9fc0ccd3068f1b64ab8b68e2a5c9d4a516/search-parts/src/components/SliderComponent.tsx#L207-L219

**Comment by jhholm** [comment]
So the easy workaround, but perhaps not the correct one would be to create handlebars helper that truncates the data from context that is not used

```
        this.Handlebars.registerHelper("truncateContext", (context: any) => {
            return {
                slots: context.slots,
                paging: context.paging,
                theme: context.theme,
                context: context.context,
                instanceId: context.instanceId,
                properties: context.properties,
                utils: context.utils
            }
        })
```

and then changing **data-context** to be applied in a template as
```
data-context="{{JSONstringify (truncateContext @root)}}"
```

**Comment by jhholm** [comment]
Could also be possible to use object destructuring in the truncation, and just remove context item
```
const { item, ...truncated } = context;
return truncated;
```

I did notice a few seconds quicker rendering with 100 items with the truncated context mentioned above.

**Comment by jhholm** [comment]
I created a branch with the changes if you want to take a look
[See changes here](https://github.com/microsoft-search/pnp-modern-search/compare/develop...jhholm:pnp-modern-search:context-truncation)

**Comment by wobba** [comment]
I like this approach and I'm ok merging it - with a release note specifying a possible breaking change if people were using weird things :)

**Comment by wobba** [comment]
Fixed with #3306
